File: src/year2024/day2/day2.go
// [Day 2: Red-Nosed Reports](https://adventofcode.com/2024/day/2)

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// isSafe returns true if the all levels are safe:
//   - The levels are either all increasing or all decreasing.
//   - Any two adjacent levels differ by at least one and at most three.
func isSafe(v []int) bool {

	decreasing, increasing := true, true

	for i := 0; i < len(v)-1; i++ {
		if !(1 <= v[i]-v[i+1] && v[i]-v[i+1] <= 3) {
			decreasing = false
		}

		if !(1 <= v[i+1]-v[i] && v[i+1]-v[i] <= 3) {
			increasing = false
		}
	}

	return decreasing || increasing
}

func main() {

	inputFile := "input.txt"
	if len(os.Args) >= 2 {
		inputFile = os.Args[1]
	}
	file, err := os.Open(inputFile)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	// Read input file
	scanner := bufio.NewScanner(file)
	var lines [][]int
	for scanner.Scan() {
		line := scanner.Text()
		numStrings := strings.Split(line, " ")

		var nums []int
		for _, str := range numStrings {
			num, _ := strconv.Atoi(str)
			nums = append(nums, num)
		}

		lines = append(lines, nums)
	}

	// Part 1
	part1 := 0
	for _, nums := range lines {
		if isSafe(nums) {
			part1 += 1
		}
	}
	fmt.Println(part1)

	// Part 2
	part2 := 0
	for _, nums := range lines {
		if isSafe(nums) {
			part2++
		} else {
			for i := 0; i < len(nums); i++ {
				temp := append([]int(nil), nums[:i]...)
				temp = append(temp, nums[i+1:]...)
				if isSafe(temp) {
					part2++
					break
				}
			}
		}
	}
	fmt.Println(part2)
}

File: src/year2024/day2/day2.rs
//! [Day 2: Red-Nosed Reports](https://adventofcode.com/2024/day/2)

struct Puzzle {
    reports: Vec<Vec<i32>>,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        let mut reports = Vec::new();

        for line in data.lines() {
            reports.push(
                line.split_whitespace()
                    .filter_map(|s: &str| s.parse().ok())
                    .collect(),
            );
        }

        Self { reports }
    }

    fn is_safe(v: &[i32]) -> bool {
        let safe_rule = |a: i32, b: i32| -> bool { (1 <= a - b) && (a - b <= 3) };

        let increasing = v.windows(2).all(|pair| safe_rule(pair[0], pair[1]));
        let decreasing = v.windows(2).all(|pair| safe_rule(pair[1], pair[0]));

        increasing || decreasing
    }

    fn is_safe_except_one(v: &[i32]) -> bool {
        for i in 0..v.len() {
            let mut w = v.to_vec();
            w.remove(i);
            if Self::is_safe(&w) {
                return true;
            }
        }
        false
    }

    /// Solve part one.
    fn part1(&self) -> usize {
        self.reports
            .iter()
            .filter(|v: &&Vec<i32>| Self::is_safe(v))
            .count()
    }

    /// Solve part two.
    fn part2(&self) -> usize {
        self.reports
            .iter()
            .filter(|v: &&Vec<i32>| Self::is_safe(v) || Self::is_safe_except_one(v))
            .count()
    }
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (usize, usize) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 2);
    }

    #[test]
    fn test02() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 4);
    }
}
